feat: Hyperwave deposit flow - Zap integration

laurentsenta opened the pull request:
Contributes to https://lifi.atlassian.net/browse/LF-14699
Sister PR: https://github.com/jumperexchange/jumper-backend/pull/391

Test page: http://localhost:3000/zap/test-zap-hyperwave

- [x] Add hyperwave chain definition from the viem PR they sent
- [x] Add a `Zapper` strategy to switch between different strategy (ones with minting or not)
- [x] Update types to sync with the backend a bit.
- [x] Implement the minting logic for hyperwave.
- [x] Add a new model where the ABI might rely on different contracts. There is a "contract" key in each function definition, and the list of contract should be defined in the "contracts" key on the market.

laurentsenta committed:
feat: draft zap hyperwave

laurentsenta committed:
feat: draft zap mint

laurentsenta committed:
feat: zap factory + contract resolution + types

laurentsenta committed:
fix: mint type is bigint

laurentsenta committed:
fix: typo in zap.interface optional contracts

admsk2 committed:
chore: Modular zaps execution;

mmioana commented on src/providers/ZapInitProvider/Zapper.ts:
```diff
@@ -56,6 +80,171 @@ export class DefaultZapper implements ZapperStrategy {
+    sendCallsExtraParams: SendCallsExtraParams,
+  ): Promise<any[]> => {
+    // Validation logic (keep this)
+    const {
+      chainId: currentChainId,
+      address: currentAddress,
+      zapData: integrationData,
+      projectData,
+    } = sendCallsExtraParams;
+
+    if (!currentChainId || !currentAddress) {
+      throw new Error('Missing chainId or address');
+    }
+
+    const currentRouteFromToken = this.currentRoute.fromToken;
+    const depositAddress = integrationData.market?.address as EVMAddress;
+    const depositToken = integrationData.market?.depositToken?.address;
+    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
+    const depositChainId = projectData.chainId;
+
+    if (!depositChainId) {
+      throw new Error('Deposit chain id is undefined.');
+    }
+
+    if (!depositAddress || !depositToken) {
+      throw new Error('Deposit address or token is undefined.');
+    }
+
+    if (!depositTokenDecimals) {
+      throw new Error('Deposit token decimals is undefined.');
+    }
+
+    // @Note this works only for EVM chains
+    const isNativeSourceToken = currentRouteFromToken.address === zeroAddress;
+
+    console.warn('Using native source token:', isNativeSourceToken);
+
+    if (isNativeSourceToken) {
+      throw new Error('Native source token is not supported.');
+    }
```
I would keep the validation logic outside separate from the `buildContractInstructions`, either in another method called before this one, or just in the `sendCalls` fn

mmioana commented on src/providers/ZapInitProvider/Zapper.ts:
What if we'd use an instruction builder so it will be easier to separate between the instructions' implementation, something along these lines
```
export interface InstructionBuilder {
  buildApproveInstruction: (
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ) => Promise<any | null>;
  
  buildDepositInstruction: (
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ) => Promise<any | null>;
  
  buildTransferInstruction: (
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ) => Promise<any | null>;
}

export class DefaultInstructionBuilder implements InstructionBuilder {
  async buildApproveInstruction(
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ): Promise<any | null> {
    const {
      address: currentAddress,
      zapData: integrationData,
      projectData,
    } = sendCallsExtraParams;

    const depositAddress = integrationData.market?.address as EVMAddress;
    const depositToken = integrationData.market?.depositToken?.address;
    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
    const depositChainId = projectData.chainId;

    if (!depositToken || !depositTokenDecimals) {
      throw new Error('Deposit token or decimals are undefined');
    }

    const constraints = [
      greaterThanOrEqualTo(parseUnits('0.1', depositTokenDecimals)),
    ];

    return await buildContractComposable(oNexus, {
      address: zapper.getApproveAddress(),
      chainId: depositChainId,
      abi: integrationData.abi.approve,
      functionName: integrationData.abi.approve.name,
      gasLimit: 100000n,
      args: [
        depositAddress,
        runtimeERC20BalanceOf({
          targetAddress: oNexus.addressOn(depositChainId, true) as EVMAddress,
          tokenAddress: depositToken,
          constraints,
        }),
      ],
    });
  }

  async buildDepositInstruction(
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ): Promise<any | null> {
    const {
      address: currentAddress,
      zapData: integrationData,
      projectData,
    } = sendCallsExtraParams;

    const depositToken = integrationData.market?.depositToken?.address;
    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
    const depositChainId = projectData.chainId;

    if (!depositToken || !depositTokenDecimals) {
      throw new Error('Deposit token or decimals are undefined');
    }

    const constraints = [
      greaterThanOrEqualTo(parseUnits('0.1', depositTokenDecimals)),
    ];

    const depositInputs = integrationData.abi.deposit.inputs;
    const depositArgs = depositInputs.map((input: AbiParameter) => {
      if (input.type === 'uint256') {
        return runtimeERC20BalanceOf({
          targetAddress: oNexus.addressOn(depositChainId, true) as EVMAddress,
          tokenAddress: depositToken,
          constraints,
        });
      } else if (input.type === 'address') {
        return currentAddress;
      }
      throw new Error(`Unsupported deposit input type: ${input.type}`);
    });

    return await buildContractComposable(oNexus, {
      address: zapper.getDepositAddress(),
      chainId: depositChainId,
      abi: integrationData.abi.deposit,
      functionName: integrationData.abi.deposit.name,
      gasLimit: 1000000n,
      args: depositArgs,
    });
  }

  async buildTransferInstruction(
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ): Promise<any | null> {
    const {
      address: currentAddress,
      zapData: integrationData,
      projectData,
    } = sendCallsExtraParams;

    const depositAddress = integrationData.market?.address as EVMAddress;
    const depositToken = integrationData.market?.depositToken?.address;
    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
    const depositChainId = projectData.chainId;

    if (!depositToken || !depositTokenDecimals) {
      throw new Error('Deposit token or decimals are undefined');
    }

    const constraints = [
      greaterThanOrEqualTo(parseUnits('0.1', depositTokenDecimals)),
    ];

    const depositInputs = integrationData.abi.deposit.inputs;
    const depositHasAddressArg = depositInputs.some(
      (input: AbiParameter) => input.type === 'address',
    );

    if (!depositHasAddressArg) {
      return await buildContractComposable(oNexus, {
        address: depositAddress,
        chainId: depositChainId,
        abi: integrationData.abi.transfer,
        functionName: integrationData.abi.transfer.name,
        gasLimit: 200000n,
        args: [
          currentAddress,
          runtimeERC20BalanceOf({
            targetAddress: oNexus.addressOn(depositChainId, true) as EVMAddress,
            tokenAddress: depositAddress,
            constraints,
          }),
        ],
      });
    }
    return null;
  }
}

export class HyperwaveInstructionBuilder extends DefaultInstructionBuilder {
  async buildDepositInstruction(
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
    zapper: ZapperStrategy,
  ): Promise<any | null> {
    // Hyperwave-specific deposit logic
    const {
      address: currentAddress,
      zapData: integrationData,
      projectData,
    } = sendCallsExtraParams;

    const depositToken = integrationData.market?.depositToken?.address;
    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
    const depositChainId = projectData.chainId;

    if (!depositToken || !depositTokenDecimals) {
      throw new Error('Deposit token or decimals are undefined');
    }

    const constraints = [
      greaterThanOrEqualTo(parseUnits('0.1', depositTokenDecimals)),
    ];

    // Hyperwave-specific minimum mint calculation
    let minimumMint: bigint | null = await zapper.computeMinimumMint();
    const depositInputs = integrationData.abi.deposit.inputs;
    const depositArgs = depositInputs.map((input: AbiParameter) => {
      if (input.type == 'uint256' && input.name === 'minimumMint') {
        if (minimumMint === null || minimumMint <= 0) {
          throw new Error('Minimum mint is not set');
        }
        return minimumMint;
      } else if (input.type === 'address') {
        return currentAddress;
      }
      throw new Error(`Unsupported deposit input type: ${input.type}`);
    });

    return await buildContractComposable(oNexus, {
      address: zapper.getDepositAddress(),
      chainId: depositChainId,
      abi: integrationData.abi.deposit,
      functionName: integrationData.abi.deposit.name,
      gasLimit: 1000000n,
      args: depositArgs,
    });
  }
}
```

And then we can adapt the strategy interface

```
export interface ZapperStrategy {
  getApproveAddress: () => `0x${string}`;
  getDepositAddress: () => `0x${string}`;
  computeMinimumMint: () => Promise<bigint | null>;
  getInstructionBuilder: () => InstructionBuilder;
  getSteps: () => string[];
  
  buildContractInstructions: (
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
  ) => Promise<any[]>;
}
```

and the implementation for the default one

```
 getSteps = (): string[] => {
    return ['approve', 'deposit', 'transfer'];
  };

  getInstructionBuilder = (): InstructionBuilder => {
    return new DefaultInstructionBuilder();
  };

  buildContractInstructions = async (
    oNexus: MultichainSmartAccount,
    sendCallsExtraParams: SendCallsExtraParams,
  ): Promise<any[]> => {
    // Validation logic (keep this)
    const {
      chainId: currentChainId,
      address: currentAddress,
      zapData: integrationData,
      projectData,
    } = sendCallsExtraParams;

    if (!currentChainId || !currentAddress) {
      throw new Error('Missing chainId or address');
    }

    const currentRouteFromToken = this.currentRoute.fromToken;
    const depositAddress = integrationData.market?.address as EVMAddress;
    const depositToken = integrationData.market?.depositToken?.address;
    const depositTokenDecimals = integrationData.market?.depositToken.decimals;
    const depositChainId = projectData.chainId;

    if (!depositChainId) {
      throw new Error('Deposit chain id is undefined.');
    }

    if (!depositAddress || !depositToken) {
      throw new Error('Deposit address or token is undefined.');
    }

    if (!depositTokenDecimals) {
      throw new Error('Deposit token decimals is undefined.');
    }

    // @Note this works only for EVM chains
    const isNativeSourceToken = currentRouteFromToken.address === zeroAddress;

    console.warn('Using native source token:', isNativeSourceToken);

    if (isNativeSourceToken) {
      throw new Error('Native source token is not supported.');
    }

    // Now use the instruction builder pattern
    const instructions: any[] = [];
    const instructionBuilder = this.getInstructionBuilder();
    
    // Execute each step using the instruction builder
    for (const step of this.getSteps()) {
      let stepInstruction: any | null = null;
      
      switch (step) {
        case 'approve':
          stepInstruction = await instructionBuilder.buildApproveInstruction(
            oNexus,
            sendCallsExtraParams,
            this,
          );
          break;
        case 'deposit':
          stepInstruction = await instructionBuilder.buildDepositInstruction(
            oNexus,
            sendCallsExtraParams,
            this,
          );
          break;
        case 'transfer':
          stepInstruction = await instructionBuilder.buildTransferInstruction(
            oNexus,
            sendCallsExtraParams,
            this,
          );
          break;
        default:
          throw new Error(`Unknown step: ${step}`);
      }
      
      if (stepInstruction) {
        instructions.push(stepInstruction);
      }
    }
    
    return instructions;
  };
```
and for hyperwave
```
getInstructionBuilder = (): InstructionBuilder => {
    return new HyperwaveInstructionBuilder();
  };
```

admsk2 committed:
chore: Validate params separately;

admsk2 committed:
Merge branch 'develop' into LF-14699-hyperwave-deposit-flow-zap-integration

mmioana committed:
chore: extract instructions building (#2112)

admsk2 committed:
chore: Hyperwave deposit updates;

admsk2 committed:
fix: Approval fix;

admsk2 committed:
Merge branch 'develop' into LF-14699-hyperwave-deposit-flow-zap-integration

laurentsenta commented on src/providers/ZapInitProvider/Zapper.ts:
```diff
@@ -51,10 +51,8 @@ export const approve: ZapInstruction = async (
-
   return await buildContractComposable(oNexus, {
-    address: zapper.getApproveAddress(),
+    address: depositToken,
```
Nice, I  missed that you call approve on the token itself with the `BoringVault` address 🤦 

Can we assume the `depositAddress` below is always the `address` we use for the token?

laurentsenta commented on src/providers/ZapInitProvider/Zapper.ts:
```diff
@@ -0,0 +1,490 @@
+  approve,
+  deposit: hyperwaveDeposit,
+  transfer,
+};
```
There's 3 level of indirection (strategy -> command -> execute), 
You could expose `computeMinimumMint` here (with a function that returns null for the default Commands) and drop the HyperwaveZapper class.

laurentsenta commented on src/providers/ZapInitProvider/Zapper.ts:
```diff
@@ -0,0 +1,490 @@
+        sendCallsExtraParams.currentRoute,
+      );
+  }
+};
```
If we're going this route, we have

`makeZapper` switches on the type to create a class (one level of indirection), that then expose different implementation of getCommands (one level of indirection), that returns different const values (another indirection).

You can get the same flexibility, while removing 2 layers, and dropping the class dispatches:

```suggestion
export const buildContractInstructions = (sendCallsExtraParams: SendCallsExtraParams) => {
  if (!sendCallsExtraParams.currentRoute) {
    throw new Error('Current route is not set');
  }

  switch (sendCallsExtraParams.projectData.project) {
    case 'hyperwave':
      return buildContractInstructions_(hyperwaveCommands, sendCallsExtraParams);
    default:
      return buildContractInstructions_(defaultCommands, sendCallsExtraParams);
  }
};
```

laurentsenta commented on src/providers/ZapInitProvider/Zapper.ts:
```diff
@@ -0,0 +1,490 @@
+    return null;
+  };
+
+  getSteps = (): string[] => {
+    return ['approve', 'deposit', 'transfer'];
+  };
+
+  getCommands = (): Record<string, ZapInstruction> => {
+    return defaultCommands;
+  };
```
Since these are tightly dependant, what if we stored them in the commands as in:

```ts
const defaultZap = {
   commands: {approve, deposit, ...}
   steps: [approve, deposit, ...]
}
```

laurentsenta replied:
Example in https://github.com/jumperexchange/jumper-exchange/blob/57bdf8e75ee59c2211c9e9fae13e8d13ea33b8b1/src/providers/ZapInitProvider/ModularZaps/HyperwaveZap.ts#L107-L114

mmioana committed:
feat: add katana chain (#2121)

mmioana committed:
chore: remove gaszip from allowed bridges

laurentsenta committed:
feat(zaps): bypass widget calls on same token tx (#2122)

Closes https://lifi.atlassian.net/browse/LF-14731
Ignore widget calls data when we're doing a same token deposit.

admsk2 committed:
Merge branch 'develop' into LF-14699-hyperwave-deposit-flow-zap-integration

laurentsenta committed:
LF- 14699: drop indirection layers (#2127)

- [x] Implements modular refactor
- [x] fixes Instruction types (We pass array of arrays)

We added a modular layer for zaps. This PR drops the genericity layer (ZapperStrategy) that is not useful anymore, and creates too many level of indirections.

mmioana reviewed (approved):
Let us 🚢
